src/serde.rs

```diff
@@ -0,0 +1,279 @@
+//! map.insert('B', 2);
+//! map.insert('C', 2);
+//!
+//! // serialize the map
+//! let json = serde_json::to_string(&map).unwrap();
+//!
+//! // deserialize it into a bimap
+//! let bimap: BiHashMap<char, i32> = serde_json::from_str(&json).unwrap();
+//!
+//! // deserialization succeeds, but the bimap is now in a non-deterministic
+//! // state - either ('B', 2) or ('C', 2) will have been overwritten while
+//! // deserializing, but this depends on the iteration order of the original
+//! // HashMap that was serialized.
+//!
+//! // we can still demonstrate that certain properties of the bimap are still
+//! // in a known state, but this shouldn't be relied upon
+//! assert_eq!(bimap.len(), 2);
+//! assert_eq!(bimap.get_by_left(&'A'), Some(&1));
+//! assert!(bimap.get_by_left(&'B') == Some(&2) || bimap.get_by_left(&'C') == Some(&2))
+//! ```
+//!
+//! The reverse is also possible: bimaps may be serialized and then
+//! deserialized as other compatible types, such as a [`HashMap`].
+//!
+//! ```
+//! # use std::collections::HashMap;
+//! # use bimap::BiHashMap;
+//! // construct a bimap
+//! let mut bimap = BiHashMap::new();
+//! 
+//! // insert some pairs
+//! bimap.insert('A', 1);
+//! bimap.insert('B', 2);
+//! bimap.insert('C', 3);
+//!
+//! // serialize the bimap
+//! let json = serde_json::to_string(&bimap).unwrap();
+//!
+//! // deserialize it as a regular map
+//! let map: HashMap<char, i32> = serde_json::from_str(&json).unwrap();
+//!
+//! // this succeeds and the result is sensible, but this is still an
+//! // implementation detail and shouldn't be relied upon.
+//! assert_eq!(map.len(), 3);
+//! assert_eq!(map[&'A'], 1);
+//! assert_eq!(map[&'B'], 2);
+//! assert_eq!(map[&'C'], 3);
+//! ```
+//! [`BiHashMap`]: crate::BiHashMap
+//! [`BiBTreeMap`]: crate::BiBTreeMap
+//! [`HashMap`]: https://doc.rust-lang.org/std/collections/struct.HashMap.html
+
+use crate::{BiHashMap, BiBTreeMap};
+use serde::{Serializer, Serialize, Deserializer, Deserialize};
+use serde::de::{Visitor, MapAccess};
+use std::hash::Hash;
+use std::fmt::{Formatter, Result as FmtResult};
+use std::marker::PhantomData;
+use std::default::Default;
+
+/// Serializer for `BiHashMap`
+impl<L, R> Serialize for BiHashMap<L, R>
+where
+    L: Serialize + Eq + Hash,
+    R: Serialize + Eq + Hash,
+{
+    fn serialize<S: Serializer>(&self, ser: S) -> Result<S::Ok, S::Error> {
+        ser.collect_map(self.iter())
+    }
+}
+
+/// Visitor to construct `BiHashMap` from serialized map entries
+struct BiHashMapVisitor<L, R> {
+    marker: PhantomData<BiHashMap<L, R>>
+}
+
+impl<'de, L, R> Visitor<'de> for BiHashMapVisitor<L, R>
+where
+    L: Deserialize<'de> + Eq + Hash,
+    R: Deserialize<'de> + Eq + Hash,
+{
+    fn expecting(&self, f: &mut Formatter) -> FmtResult {
+        write!(f, "a map")
+    }
+
+    type Value = BiHashMap<L, R>;
+    fn visit_map<A: MapAccess<'de>>(self, mut entries: A) -> Result<Self::Value, A::Error> {
+        let mut map = match entries.size_hint() {
```

Comment:
nice use of `size_hint`  - going to use something similar for `FromIterator` as well.